src/SharpBrick.PoweredUp.BlueGigaBLE/BlueGigaBLEPoweredUpBluetoothCharacteristic.cs

```diff
@@ -0,0 +1,136 @@
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Bluegiga.BLE.Events.ATTClient;
+using Bluegiga.BLE.Events.Attributes;
+using Microsoft.Extensions.Logging;
+using SharpBrick.PoweredUp.Bluetooth;
+
+namespace SharpBrick.PoweredUp.BlueGigaBLE
+{
+    public class BlueGigaBLEPoweredUpBluetoothCharacteristic : IPoweredUpBluetoothCharacteristic, IDisposable, IBlueGigaLogger
+    {
+        #region Properties
+        /// <summary>
+        /// The Service this Characteristic belongs to; will be set in the calling/constructing Service
+        /// </summary>
+        public BlueGigaBLEPoweredUpBluetoothService myService { get; set; }
+        /// <summary>
+        /// The connectionof the device offering this characteristic
+        /// </summary>
+        public byte connection { get { return myService.myDevice.ConnectionHandle; } }
+        /// <summary>
+        /// The characteristic-handle on the connection which has to be used by the bluegiga-adapter to adress this character
+        /// </summary>
+        public ushort myhandle { get; set; }
+        /// <summary>
+        /// The Logger for this object
+        /// </summary>
+        public readonly ILogger _logger;
+        #endregion
+        #region Constructors
+        public BlueGigaBLEPoweredUpBluetoothCharacteristic(BlueGigaBLEPoweredUpBluetoothService myService, ushort myhandle, Guid characteristicUuid, ILogger logger, bool traceDebug)
+        {
+            this.myService = myService ?? throw new ArgumentNullException(nameof(myService));
+            this.myhandle = myhandle;
+            Uuid = characteristicUuid;
+            this._logger = logger;
+            this.TraceDebug = traceDebug;
+        }
+        #endregion
+        #region IPoweredUpBlueToothCharacteristic
+        /// <summary>
+        /// The UUID of the characteristic
+        /// </summary>
+        public Guid Uuid { get; }
+        public async Task<bool> NotifyValueChangeAsync(Func<byte[], Task> notificationHandler)
+        {
+            if (notificationHandler is null)
+            {
+                throw new ArgumentNullException(nameof(notificationHandler));
+            }
+            this.myService.myDevice.Bglib.BLEEventATTClientAttributeValue += new Bluegiga.BLE.Events.ATTClient.AttributeValueEventHandler(AttributeValueChangedEventHandler);
+            void AttributeValueChangedEventHandler(object sender, AttributeValueEventArgs e)
+            {
+                
+                if ((e.atthandle == this.myhandle) && (e.connection == this.connection))
+                {
+                    notificationHandler( e.value);
+                }
+            }
+            Byte[] cmd = this.myService.myDevice.Bglib.BLECommandATTClientWriteCommand(connection , 0x0F , new byte[] { 0x01, 0x00 });
+            var status = await Task.Run( () => this.myService.myDevice.Bglib.SendCommand( this.myService.myDevice.myBluetoothAdapter.SerialAPI,  cmd));
```

Comment:
you work here asynchronously ... but not in other places. which can be fine, just, is it intentional?

Reply:
I made it now async/await wherever possible/reasonable for me. Please review...

Reply:
and surely I did a lot of testing afterwards (deadlock,...)